reports/cfgmgmnt.go

```diff
@@ -0,0 +1,71 @@
+package reports
+
+import (
+	"encoding/json"
+	"os/exec"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// Facts stores output of puppet & chef facts command
+type Facts map[string]interface{}
+
+// GetFacts will exec into the Facts interface
+func GetFacts(confTool, cmdPath, cmdArgs string) (Facts, error) {
+	args := strings.Split(cmdArgs, " ")
+	cmd := exec.Command(cmdPath, args...)
+
+	o, err := cmd.Output()
+	if err != nil {
+		return Facts{}, errors.Wrap(err, "exec puppet facts")
+	}
+
+	var facts Facts
+
+	if err := json.Unmarshal(o, &facts); err != nil {
+		return Facts{}, errors.Wrap(err, "failed unmarshalling Puppet Facts")
+	}
+
+	switch confTool {
+	case "puppet":
+		facts = facts["values"].(map[string]interface{})
+	case "salt":
+		facts = facts["local"].(map[string]interface{})
+	}
+
+	stripEmpty(facts)
+	facts = FlattenMapStringInterface(facts)
+
+	return facts, nil
+}
+
+// FlattenMapStringInterface takes a map[string]interface and returns a flattened map[string]interface{}
+func FlattenMapStringInterface(m map[string]interface{}) map[string]interface{} {
+	o := make(map[string]interface{})
+	for k, v := range m {
+		switch child := v.(type) {
+		case map[string]interface{}:
+			nm := FlattenMapStringInterface(child)
+			for nk, nv := range nm {
+				o[k+"=>"+nk] = nv
+			}
+		default:
+			o[k] = v
+		}
+	}
+	return o
+}
+
+// stripEmpty will remove null values from a map recursively
+func stripEmpty(aMap map[string]interface{}) {
+	for key, val := range aMap {
+		if val == nil {
+			delete(aMap, key)
+		}
```

Comment:
I could move this delete into the FlattenMapStringInterface method. These two loops are doing  essentially the same thing. I’m sure that would be more efficient.

Reply:
Agree about that